src/content/guides/asset-modules.md

```diff
@@ -0,0 +1,353 @@
+      }
+    ]
+  }
+};
+```
+
+__src/index.js__
+
+``` diff
+- import mainImage from './images/main.png';
++ import metroMap from './images/matro.svg';
+
+// ...
+
+- img.src = mainImage; // '/dist/151cfcfa1bd74779aadb.png'
++ block.style.background = `url(${metroMap}); // url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDo...vc3ZnPgo=)
+```
+
+All `svg` files will be injected into the bundles as data-url. Just like with `url-loader`
+
+### Custom data-url generator
+
+By default, data-url is a file content encoded with base64.
+
+If you want to use a custom encoding algorithm, you may specify a custom function to encode a file content:
+
+__webpack.config.js__
+
+``` diff
+const path = require('path');
++ const svgToMiniDataURI = require('mini-svg-data-uri');
+
+module.exports = {
+  entry: './src/index.js',
+  output: {
+    filename: 'main.js',
+    path: path.resolve(__dirname, 'dist')
+  },
+  experiments: {
+    asset: true
+  },
+  module: {
+    rules: [
+      {
+        test: /\.svg/,
+        type: 'asset/inline',
++       generator: {
++         dataUrl: content => {
++           content = content.toString();
++           return svgToMiniDataURI(content);
++         }
++       }
+      }
+    ]
+  }
+};
+```
+
+This one will use `mini-svg-data-uri` package to encode all `svg` files.
+
+## Source assets
+
+__webpack.config.js__
+
+``` diff
+const path = require('path');
+- const svgToMiniDataURI = require('mini-svg-data-uri');
+
+module.exports = {
+  entry: './src/index.js',
+  output: {
+    filename: 'main.js',
+    path: path.resolve(__dirname, 'dist')
+  },
+  experiments: {
+    asset: true
+  },
+  module: {
+    rules: [
+      {
+-       test: /\.svg/,
+-       type: 'asset/inline',
+-       generator: {
+-         dataUrl: content => {
+-           if (typeof content !== "string") {
+-             content = content.toString();
+-           }
+-
+-           return svgToMiniDataURI(content);
+-         }
+-       }
++       test: /\.txt/,
++       type: 'asset/source',
+      }
+    ]
+  }
+};
+```
+
+__src/example.txt__
+
+``` text
+Hello world
+```
+
+__src/index.js__
+
+``` diff
+- import metroMap from './images/matro.svg';
++ import examleText from './example.txt';
+
+// ...
+
+- block.style.background = `url(${metroMap}); // url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDo...vc3ZnPgo=)
++ block.textContent = examleText; // 'Hello world'
+```
+
+All `txt` files will be injected into the bundles as is. Just like with `raw-loader`
+
+## General asset type
+
+__webpack.config.js__
+
+``` diff
+const path = require('path');
+
+module.exports = {
+  entry: './src/index.js',
+  output: {
+    filename: 'main.js',
+    path: path.resolve(__dirname, 'dist')
+  },
+  experiments: {
+    asset: true
+  },
+  module: {
+    rules: [
+      {
++       test: /\.txt/,
++       type: 'asset',
+      }
+    ]
+  }
+};
+```
+
+This one will automatically chooses between `resource` and `inline` by a condition.
```

Comment:
```suggestion
Now webpack will automatically choose between `resource` and `inline` by following a default condition: a file with size less than 8kb will be treated as a `inline` module type and `resource` module type otherwise.
```